agent/stats/engine.go

```diff
@@ -0,0 +1,446 @@
+func (resolver *DockerContainerMetadataResolver) ResolveTask(dockerID string) (*api.Task, error) {
+	if resolver.dockerTaskEngine == nil {
+		return nil, errors.New("Docker task engine uninitialized.")
+	}
+	task, found := resolver.dockerTaskEngine.State().TaskById(dockerID)
+	if !found {
+		return nil, errors.New("Could not map docker id to task")
+	}
+
+	return task, nil
+}
+
+// ResolveName resolves the container name, given container id.
+func (resolver *DockerContainerMetadataResolver) ResolveName(dockerID string) (string, error) {
+	if resolver.dockerTaskEngine == nil {
+		return "", errors.New("Docker task engine uninitialized.")
+	}
+	container, found := resolver.dockerTaskEngine.State().ContainerById(dockerID)
+	if !found {
+		return "", errors.New("Could not map docker id to container")
+	}
+
+	return container.DockerName, nil
+}
+
+// NewDockerStatsEngine creates a new instance of the DockerStatsEngine object.
+// MustInit() must be called to initialize the fields of the new event listener.
+func NewDockerStatsEngine() *DockerStatsEngine {
+	if dockerStatsEngine == nil {
+		dockerStatsEngine = &DockerStatsEngine{
+			client:             nil,
+			resolver:           nil,
+			tasksToContainers:  make(map[string]map[string]*CronContainer),
+			tasksToDefinitions: make(map[string]*taskDefinition),
+		}
+	}
+
+	return dockerStatsEngine
+}
+
+// MustInit initializes fields of the DockerStatsEngine object.
+func (engine *DockerStatsEngine) MustInit(taskEngine ecsengine.TaskEngine, md *ecstcs.MetricsMetadata) error {
+	log.Info("Initializing stats engine")
+	err := engine.initDockerClient()
+	if err != nil {
+		return err
+	}
+
+	engine.metricsMetadata = md
+
+	engine.resolver, err = newDockerContainerMetadataResolver(taskEngine)
+	if err != nil {
+		return err
+	}
+
+	return engine.Init()
+}
+
+// Init initializes the docker client's event engine. This must be called
+// to subscribe to the docker's event stream.
+func (engine *DockerStatsEngine) Init() error {
+	engine.ctx, engine.unsubscribeContainerEvents = context.WithCancel(context.TODO())
+	err := engine.openEventStream()
+	if err != nil {
+		return err
+	}
+
+	go engine.listContainersAndStartEventHandler()
+	return nil
+}
+
+// listContainersAndStartEventHandler adds existing containers to the watch-list
+// and starts the docker event handler.
+func (engine *DockerStatsEngine) listContainersAndStartEventHandler() {
+	// List and add existing containers to the list of containers to watch.
+	err := engine.addExistingContainers()
+	if err != nil {
+		log.Warn("Error listing existing containers", "err", err)
+		// Cancel the context to unsubscribe container events.
+		engine.unsubscribeContainerEvents()
+		return
+	}
+
+	go engine.handleDockerEvents()
+}
+
+// AddContainer adds a container to the map of containers being watched.
+// It also starts the periodic usage data collection for the container.
+func (engine *DockerStatsEngine) AddContainer(dockerID string) {
```

Comment:
I'm looking and trying to see where `AddContainer` is called outside the stats engine.  Is it public because you're planning to add a call here in a future commit?